{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Summary\n",
    "\n",
    "\n",
    "- There are many many duplicates produced by the COSMIC SQL query, not sure why...\n",
    "\n",
    "- 16 GB RAM is _just_ enough to run this notebook...\n",
    "\n",
    "-----"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2016-07-10 20:22:28.876320\n"
     ]
    }
   ],
   "source": [
    "%run imports.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "NOTEBOOK_NAME = 'core_load_data'\n",
    "os.environ['NOTEBOOK_NAME'] = NOTEBOOK_NAME"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DATASETS = [\n",
    "    'alascan_gpk',\n",
    "    'curatedprotherm',\n",
    "    'guerois',\n",
    "    'kellogg',\n",
    "    'potapov',\n",
    "    'protherm',\n",
    "    'taipale',\n",
    "    'humsavar',\n",
    "    'clinvar',\n",
    "    'cosmic',\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Construct SQL query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "code_folding": [],
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# db_remote\n",
    "db_remote = datapkg.MySQL(\n",
    "    connection_string=os.environ['DATAPKG_CONNECTION_STR'] + '/elaspic', \n",
    "    shared_folder=os.environ['NOTEBOOK_NAME'], \n",
    "    storage_host=None, \n",
    "    echo=False, \n",
    "    db_engine='MyISAM')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "code_folding": [],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Combine experimental mutation sets with ELASPIC features\n",
    "sql_query_template = \"\"\"\\\n",
    "SELECT *\n",
    "\n",
    "FROM ({subquery}) p\n",
    "\n",
    "JOIN {uniprot_sequence_subquery}\n",
    "JOIN {elaspic_schema}.uniprot_domain d using (uniprot_id)\n",
    "LEFT JOIN {elaspic_schema}.uniprot_domain_template t using (uniprot_domain_id)\n",
    "LEFT JOIN {elaspic_schema}.uniprot_domain_model m using (uniprot_domain_id)\n",
    "LEFT JOIN {elaspic_schema}.uniprot_domain_mutation mut using (uniprot_id, uniprot_domain_id, mutation)\n",
    "\n",
    "-- These two lines control whether we include mutations which fall outside domain\n",
    "WHERE (\n",
    "    (model_domain_def is not NULL and elaspic.MUTATION_IN_DOMAIN(mutation, model_domain_def)) OR \n",
    "    (model_domain_def is NULL and model_errors is NULL and elaspic.MUTATION_IN_DOMAIN(mutation, domain_def))\n",
    ") \n",
    "AND (mutation REGEXP '^[A-Za-z][0-9]+[A-Za-z]$') \n",
    "AND (SUBSTRING(mutation, 1, 1) != SUBSTRING(mutation, -1))\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "code_folding": [],
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Load dataset queries\n",
    "subqueries = dict()\n",
    "\n",
    "subqueries['protein_folding_energy'] = \"\"\"\n",
    "SELECT\n",
    "    uniprot_id uniprot_id,\n",
    "    uniprot_mutation mutation,\n",
    "    ddg_exp,\n",
    "    null del_score_exp,\n",
    "    null del_class_exp\n",
    "FROM protein_folding_energy.{table_name}\n",
    "WHERE ddg_exp is not null  -- remove mutations with no experimental ddG\n",
    "-- AND (uniprot_mutation REGEXP '^[A-Za-z][0-9]+[A-Za-z]$')  -- remove weird mutations\n",
    "-- AND (SUBSTRING(uniprot_mutation, 1, 1) != SUBSTRING(uniprot_mutation, -1))  -- remove synonymous mutations\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "subqueries['humsavar'] = \"\"\"\n",
    "SELECT\n",
    "    uniprot_id uniprot_id,\n",
    "    uniprot_mutation mutation,\n",
    "    null ddg_exp,\n",
    "    null del_score_exp,\n",
    "    CASE type_of_variant WHEN 'Polymorphism' THEN 0 WHEN 'Disease' then 1 ELSE NULL END del_class_exp\n",
    "FROM uniprot.humsavar\n",
    "WHERE type_of_variant = 'Polymorphism' OR type_of_variant = 'Disease'\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "subqueries['clinvar'] = \"\"\"\n",
    "SELECT\n",
    "    v.mutationassessor_uniprotid uniprot_name,\n",
    "    v.mutationassessor_variant mutation,\n",
    "    null ddg_exp,\n",
    "    v.mutationassessor_score del_score_exp,\n",
    "    0 del_class_exp\n",
    "FROM clinvar_local.clinvar_benign c\n",
    "JOIN dbnsfp.variant v ON (c.id = v.rs_dbsnp146)\n",
    "WHERE v.mutationassessor_uniprotid IS NOT NULL AND v.mutationassessor_variant IS NOT NULL\n",
    "\n",
    "    UNION ALL\n",
    "\n",
    "SELECT \n",
    "    v.mutationassessor_uniprotid uniprot_name,\n",
    "    v.mutationassessor_variant mutation,\n",
    "    null ddg_exp,\n",
    "    v.mutationassessor_score del_score_exp,\n",
    "    1 del_class_exp\n",
    "FROM clinvar_local.clinvar c\n",
    "JOIN dbnsfp.variant v ON (c.id = v.rs_dbsnp146)\n",
    "WHERE v.mutationassessor_uniprotid IS NOT NULL AND v.mutationassessor_variant IS NOT NULL\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "subqueries['cosmic'] = \"\"\"\n",
    "SELECT \n",
    "    v.mutationassessor_uniprotid uniprot_name,\n",
    "    v.mutationassessor_variant mutation,\n",
    "    NULL ddg_exp,\n",
    "    cme.fathmm_score del_score_exp,\n",
    "    CASE cme.fathmm_prediction WHEN 'NEUTRAL' THEN 0 WHEN 'PATHOGENIC' THEN 1 ELSE NULL END del_class_exp\n",
    "FROM cosmic_new.cosmic_coding_muts ccm \n",
    "JOIN cosmic_new.cosmic_mutant_export cme ON (cme.mutation_id = ccm.id)\n",
    "JOIN dbnsfp.variant v ON (ccm.chrom = v.chr AND ccm.pos = v.pos_1based AND ccm.ref = v.ref AND ccm.alt = v.alt)\n",
    "WHERE cme.fathmm_score IS NOT NULL AND cme.fathmm_prediction IS NOT NULL\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "assert not any((';' in v) for v in subqueries.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "code_folding": [],
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Load dataset functions\n",
    "def load_dataset(table_name, elaspic_schema, engine=None):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    if engine is None:\n",
    "        engine = sa.create_engine(os.environ['DATAPKG_CONNECTION_STR'] + '/elaspic')\n",
    "\n",
    "    # Create sql query\n",
    "    if table_name in ['humsavar', 'clinvar', 'cosmic']:\n",
    "        subquery = subqueries[table_name]\n",
    "    else:\n",
    "        subquery = subqueries['protein_folding_energy'].format(table_name=table_name)\n",
    "\n",
    "    if table_name in ['clinvar', 'cosmic']:\n",
    "        # Do not join on a table with splice variants if using `uniprot_name`\n",
    "        uniprot_sequence_subquery = \"\"\"\\\n",
    "(SELECT uniprot_acc uniprot_id, uniprot_id uniprot_name, uniprot_sequence \n",
    " FROM uniprot_kb_proteomes.UP000005640_9606_fasta) us USING (uniprot_name) \\\n",
    "\"\"\"\n",
    "    else:\n",
    "        uniprot_sequence_subquery = \"\"\"\\\n",
    "uniprot_kb.uniprot_sequence us USING (uniprot_id) \\\n",
    "\"\"\"\n",
    "\n",
    "    sql_query = sql_query_template.format(\n",
    "            elaspic_schema=elaspic_schema,\n",
    "            subquery=subquery, \n",
    "            uniprot_sequence_subquery=uniprot_sequence_subquery)\n",
    "\n",
    "#     cache_file = op.join(NOTEBOOK_NAME, table_name + '.tsv.gz')\n",
    "#     if op.isfile(cache_file):\n",
    "#         logger.info(\"Reading cache file '{}'...\".format(cache_file))\n",
    "#         sql_query += ' LIMIT 0'\n",
    "#         df_header = pd.read_sql_query(sql_query, engine)\n",
    "#         df = pd.read_csv(cache_file, sep='\\t', na_values=['\\\\N'], names=df_header.columns)\n",
    "#     else:\n",
    "    df = pd.read_sql_query(sql_query, engine)\n",
    "    \n",
    "    # Read sql query\n",
    "    df = df.rename(columns={'mutation': 'uniprot_mutation'})\n",
    "    # df = ascommon.df_tools.remove_duplicate_columns(df)\n",
    "    print(\"{} [{}]: ({})\".format(table_name, elaspic_schema, df.shape[0]))\n",
    "    \n",
    "    # === All these rules could be converted to SQL... ===\n",
    "    # Remove rows with null mutations\n",
    "    null_mutants_ = (df['uniprot_mutation'].isnull()) | (df['uniprot_mutation'] == '-')\n",
    "    print(\"Removing {} null mutants...\".format(null_mutants_.sum()))\n",
    "    df = df[~null_mutants_]\n",
    "\n",
    "    # Remove SIFTS errors\n",
    "    sifts_errors_ = df['uniprot_mutation'] == '?'\n",
    "    print(\"Removing {} sifts errors...\".format(sifts_errors_.sum()))\n",
    "    df = df[~sifts_errors_]\n",
    "\n",
    "    # Remove rows with weird mutations\n",
    "    multi_mutants_ = df['uniprot_mutation'].str.contains(',')\n",
    "    print(\"Removing {} multi mutants...\".format(multi_mutants_.sum()))\n",
    "    df = df[~multi_mutants_]\n",
    "\n",
    "    # Remove wild-type rows\n",
    "    wild_ = (df['uniprot_mutation'].str.lower() == 'wild')\n",
    "    print(\"Removing {} wild mutants...\".format(wild_.sum()))\n",
    "    df = df[~wild_]\n",
    "\n",
    "    # Remove synonymous mutations\n",
    "    synonymous_ = df['uniprot_mutation'].str[0] == df['uniprot_mutation'].str[-1]\n",
    "    print(\"Removing {} synonymous mutants...\".format(synonymous_.sum()))\n",
    "    df = df[~synonymous_]\n",
    "\n",
    "    # Remove sequence mismatch mutations\n",
    "    mutation_matches_sequence_ = (\n",
    "        df[['uniprot_mutation', 'uniprot_sequence']]\n",
    "        .apply(lambda x: ascommon.sequence_tools.mutation_matches_sequence(*x), axis=1)\n",
    "    )\n",
    "    mutation_matches_sequence_ = mutation_matches_sequence_.fillna(False)\n",
    "    print(\"Removing {} mutations not matching sequence...\".format((~mutation_matches_sequence_).sum()))\n",
    "    df = df[mutation_matches_sequence_]\n",
    "\n",
    "    # Convert ELASPIC features to expected format\n",
    "    shape_before = df.shape[0]\n",
    "    df = elaspic.elaspic_predictor.format_mutation_features(df, 'core')\n",
    "    df = elaspic.elaspic_predictor.convert_features_to_differences(df)\n",
    "    shape_after = df.shape[0]\n",
    "    assert shape_before == shape_after\n",
    "    \n",
    "    # DONE!\n",
    "    print('-' * 80, flush=True)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run SQL query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "os.makedirs(NOTEBOOK_NAME, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "logger.setLevel(logging.DEBUG)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "code_folding": [],
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "alascan_gpk [elaspic]: (743)\n",
      "Removing 0 null mutants...\n",
      "kellogg [elaspic]: (1109)\n",
      "guerois [elaspic]: (903)\n",
      "Removing 0 null mutants...\n",
      "Removing 0 null mutants...\n",
      "Removing 0 sifts errors...\n",
      "Removing 0 sifts errors...\n",
      "Removing 0 sifts errors...\n",
      "Removing 0 multi mutants...\n",
      "Removing 0 multi mutants...\n",
      "Removing 0 multi mutants...\n",
      "Removing 0 wild mutants...\n",
      "Removing 0 wild mutants...\n",
      "Removing 0 wild mutants...\n",
      "Removing 0 synonymous mutants...\n",
      "Removing 0 synonymous mutants...\n",
      "Removing 0 synonymous mutants...\n",
      "Removing 1 mutations not matching sequence...\n",
      "Removing 1 mutations not matching sequence...\n",
      "Removing 3 mutations not matching sequence...\n",
      "protherm [elaspic]: (5682)\n",
      "kellogg [elaspic_training]: (2830)\n",
      "taipale [elaspic]: (1920)\n",
      "alascan_gpk [elaspic_training]: (2260)\n",
      "Removing 0 null mutants...\n",
      "Removing 0 null mutants...\n",
      "Removing 0 null mutants...\n",
      "Removing 0 null mutants...\n",
      "Removing 0 sifts errors...\n",
      "protherm [elaspic_training]: (17127)\n",
      "Removing 0 sifts errors...\n",
      "Removing 0 sifts errors...\n",
      "taipale [elaspic_training]: (4587)\n",
      "Removing 0 multi mutants...\n",
      "Removing 0 multi mutants...\n",
      "potapov [elaspic]: (2048)\n",
      "Removing 0 sifts errors...\n",
      "curatedprotherm [elaspic]: (2791)\n",
      "Removing 0 multi mutants...\n",
      "Removing 0 null mutants...\n",
      "Removing 0 multi mutants...\n",
      "curatedprotherm [elaspic_training]: (8139)\n",
      "Removing 0 null mutants...\n",
      "Removing 0 wild mutants...\n",
      "Removing 0 null mutants...\n",
      "potapov [elaspic_training]: (6141)\n",
      "Removing 0 null mutants...\n",
      "Removing 0 null mutants...\n",
      "Removing 0 wild mutants...\n",
      "Removing 0 sifts errors...\n",
      "guerois [elaspic_training]: (2374)\n",
      "Removing 0 sifts errors...\n",
      "Removing 0 sifts errors...\n",
      "Removing 0 null mutants...\n",
      "Removing 0 wild mutants...\n",
      "Removing 0 multi mutants...\n",
      "Removing 0 multi mutants...\n",
      "Removing 0 sifts errors...\n",
      "Removing 0 sifts errors...\n",
      "Removing 0 multi mutants...\n",
      "Removing 0 null mutants...\n",
      "Removing 0 synonymous mutants...\n",
      "Removing 0 wild mutants...\n",
      "Removing 0 multi mutants...\n",
      "Removing 0 multi mutants...\n",
      "Removing 0 sifts errors...\n",
      "Removing 0 synonymous mutants...\n",
      "Removing 0 wild mutants...\n",
      "Removing 0 synonymous mutants...\n",
      "Removing 0 multi mutants...\n",
      "Removing 0 wild mutants...\n",
      "Removing 0 sifts errors...\n",
      "Removing 10 mutations not matching sequence...\n",
      "Removing 0 wild mutants...\n",
      "Removing 0 wild mutants...\n",
      "Removing 0 wild mutants...\n",
      "Removing 10 mutations not matching sequence...\n",
      "Removing 0 multi mutants...\n",
      "Removing 0 synonymous mutants...\n",
      "Removing 2 mutations not matching sequence...\n",
      "Removing 0 wild mutants...\n",
      "Removing 0 synonymous mutants...\n",
      "Removing 0 synonymous mutants...\n",
      "Removing 0 synonymous mutants...\n",
      "Removing 0 wild mutants...\n",
      "Removing 0 synonymous mutants...\n",
      "Removing 0 synonymous mutants...\n",
      "Removing 0 mutations not matching sequence...\n",
      "Removing 3 mutations not matching sequence...\n",
      "Removing 0 synonymous mutants...\n",
      "Removing 0 mutations not matching sequence...\n",
      "Removing 0 synonymous mutants...\n",
      "Removing 10 mutations not matching sequence...\n",
      "Removing 14 mutations not matching sequence...\n",
      "Removing 34 mutations not matching sequence...\n",
      "Removing 4 mutations not matching sequence...\n",
      "Removing 2 mutations not matching sequence...\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "humsavar [elaspic]: (33160)\n",
      "Removing 0 null mutants...\n",
      "Removing 0 sifts errors...\n",
      "Removing 0 multi mutants...\n",
      "Removing 0 wild mutants...\n",
      "Removing 0 synonymous mutants...\n",
      "Removing 231 mutations not matching sequence...\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "cosmic [elaspic]: (788284)\n",
      "Removing 0 null mutants...\n",
      "Removing 0 sifts errors...\n",
      "Removing 0 multi mutants...\n",
      "Removing 0 wild mutants...\n",
      "Removing 0 synonymous mutants...\n",
      "Removing 3225 mutations not matching sequence...\n",
      "clinvar [elaspic]: (71013)\n",
      "Removing 0 null mutants...\n",
      "Removing 0 sifts errors...\n",
      "Removing 0 multi mutants...\n",
      "Removing 0 wild mutants...\n",
      "Removing 0 synonymous mutants...\n",
      "Removing 502 mutations not matching sequence...\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# DATA\n",
    "params = []\n",
    "for table_name in DATASETS:\n",
    "    # DATA[table_name] = load_dataset(table_name, 'elaspic', db.engine)\n",
    "    params.append((table_name, table_name, 'elaspic'))\n",
    "    if table_name not in ['humsavar', 'clinvar', 'cosmic']:\n",
    "        # DATA[table_name + '_diffseqi'] = load_dataset(table_name, 'elaspic_training', db.engine)\n",
    "        params.append((table_name + '_diffseqi', table_name, 'elaspic_training_core'))\n",
    "\n",
    "# Get data from database\n",
    "import concurrent.futures\n",
    "with concurrent.futures.ThreadPoolExecutor(len(params)) as p:\n",
    "    results = p.map(lambda x: load_dataset(*x[1:]), params)\n",
    "\n",
    "# Combine into one dict\n",
    "DATA = {p[0]: r for p, r in zip(params, results)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "with open(op.join(NOTEBOOK_NAME, 'DATA.pkl'), 'wb') as ofh:\n",
    "    pickle.dump(DATA, ofh, pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DATA_DF\n",
    "\n",
    "Combine data into a DataFrame.\n",
    "\n",
    "All $\\Delta \\Delta G$ datasets go into a single \"protherm\" dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "with open(op.join(NOTEBOOK_NAME, 'DATA.pkl'), 'rb') as ifh:\n",
    "    DATA = pickle.load(ifh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "run_control": {
     "marked": false
    }
   },
   "outputs": [],
   "source": [
    "# In case of duplicates, the dataset occuring FIRST has the priority\n",
    "df_list = []\n",
    "dataset_map = {\n",
    "    # Best sequence identity\n",
    "    'alascan_gpk': 'protherm',\n",
    "    'curatedprotherm': 'protherm',\n",
    "    'guerois': 'protherm',\n",
    "    'kellogg': 'protherm',\n",
    "    'potapov': 'protherm',\n",
    "    'protherm': 'protherm',\n",
    "    #\n",
    "    'taipale': 'taipale',\n",
    "    #\n",
    "    'humsavar': 'humsavar',\n",
    "    'clinvar': 'clinvar',\n",
    "    'cosmic': 'cosmic',\n",
    "}\n",
    "\n",
    "def append_to_df_list(key, suffix):\n",
    "    df = DATA[key + suffix]\n",
    "    df['dataset'] = dataset_map[key] + suffix\n",
    "    df_list.append(df)\n",
    "\n",
    "for key in DATASETS:\n",
    "    if key not in ['humsavar', 'clinvar', 'cosmic']:\n",
    "        suffixes = ['', '_diffseqi']\n",
    "    else:\n",
    "        suffixes = ['']\n",
    "    for suffix in suffixes:\n",
    "        append_to_df_list(key, suffix)\n",
    "\n",
    "DATA_DF = pd.concat(df_list, ignore_index=True)\n",
    "\n",
    "assert DATA_DF.shape[0] == sum(df.shape[0] for df in DATA.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for c in ['ddg_exp', 'del_score_exp', 'del_class_exp']:\n",
    "    DATA_DF[c] = DATA_DF[c].astype(float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Set 'max_seq_identity' where it is missing\n",
    "DATA_DF.loc[DATA_DF['max_seq_identity'].isnull(), 'max_seq_identity'] = (\n",
    "    DATA_DF[DATA_DF['max_seq_identity'].isnull()]\n",
    "    ['alignment_identity']\n",
    "    .apply(local.get_max_seq_identity)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>alignment_coverage</th>\n",
       "      <th>alignment_def</th>\n",
       "      <th>alignment_filename</th>\n",
       "      <th>alignment_identity</th>\n",
       "      <th>alignment_score</th>\n",
       "      <th>alignment_subdefs</th>\n",
       "      <th>backbone_clash_change</th>\n",
       "      <th>backbone_clash_wt</th>\n",
       "      <th>backbone_hbond_change</th>\n",
       "      <th>backbone_hbond_wt</th>\n",
       "      <th>cath_id</th>\n",
       "      <th>chain</th>\n",
       "      <th>chain_modeller</th>\n",
       "      <th>cis_bond_change</th>\n",
       "      <th>cis_bond_wt</th>\n",
       "      <th>dataset</th>\n",
       "      <th>db</th>\n",
       "      <th>ddg</th>\n",
       "      <th>ddg_exp</th>\n",
       "      <th>del_class_exp</th>\n",
       "      <th>del_score_exp</th>\n",
       "      <th>dg_change</th>\n",
       "      <th>dg_wt</th>\n",
       "      <th>disulfide_change</th>\n",
       "      <th>disulfide_wt</th>\n",
       "      <th>domain_def</th>\n",
       "      <th>domain_end</th>\n",
       "      <th>domain_start</th>\n",
       "      <th>electrostatic_kon_change</th>\n",
       "      <th>electrostatic_kon_wt</th>\n",
       "      <th>electrostatics_change</th>\n",
       "      <th>electrostatics_wt</th>\n",
       "      <th>energy_ionisation_change</th>\n",
       "      <th>energy_ionisation_wt</th>\n",
       "      <th>entropy_complex_change</th>\n",
       "      <th>entropy_complex_wt</th>\n",
       "      <th>entropy_mainchain_change</th>\n",
       "      <th>entropy_mainchain_wt</th>\n",
       "      <th>entropy_sidechain_change</th>\n",
       "      <th>entropy_sidechain_wt</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>helix_dipole_change</th>\n",
       "      <th>helix_dipole_wt</th>\n",
       "      <th>m_date_modified</th>\n",
       "      <th>matrix_score</th>\n",
       "      <th>max_seq_identity</th>\n",
       "      <th>mloop_entropy_change</th>\n",
       "      <th>mloop_entropy_wt</th>\n",
       "      <th>model_domain_def</th>\n",
       "      <th>model_errors</th>\n",
       "      <th>model_filename</th>\n",
       "      <th>model_filename_mut</th>\n",
       "      <th>model_filename_wt</th>\n",
       "      <th>mut_date_modified</th>\n",
       "      <th>mutation_errors</th>\n",
       "      <th>mutation_modeller</th>\n",
       "      <th>norm_dope</th>\n",
       "      <th>number_of_residues</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>partial_covalent_bonds_change</th>\n",
       "      <th>partial_covalent_bonds_wt</th>\n",
       "      <th>path_to_data</th>\n",
       "      <th>pcv_hbond_change</th>\n",
       "      <th>pcv_hbond_self_change</th>\n",
       "      <th>pcv_hbond_self_wt</th>\n",
       "      <th>pcv_hbond_wt</th>\n",
       "      <th>pcv_salt_equal_change</th>\n",
       "      <th>pcv_salt_equal_self_change</th>\n",
       "      <th>pcv_salt_equal_self_wt</th>\n",
       "      <th>pcv_salt_equal_wt</th>\n",
       "      <th>pcv_salt_opposite_change</th>\n",
       "      <th>pcv_salt_opposite_self_change</th>\n",
       "      <th>pcv_salt_opposite_self_wt</th>\n",
       "      <th>pcv_salt_opposite_wt</th>\n",
       "      <th>pcv_vdw_change</th>\n",
       "      <th>pcv_vdw_self_change</th>\n",
       "      <th>pcv_vdw_self_wt</th>\n",
       "      <th>pcv_vdw_wt</th>\n",
       "      <th>pdbfam_idx</th>\n",
       "      <th>pdbfam_name</th>\n",
       "      <th>pfam_clan</th>\n",
       "      <th>pfam_names</th>\n",
       "      <th>protein_existence</th>\n",
       "      <th>protein_name</th>\n",
       "      <th>provean_score</th>\n",
       "      <th>sasa_score</th>\n",
       "      <th>secondary_structure_change</th>\n",
       "      <th>secondary_structure_wt</th>\n",
       "      <th>sequence_version</th>\n",
       "      <th>sidechain_hbond_change</th>\n",
       "      <th>sidechain_hbond_wt</th>\n",
       "      <th>sloop_entropy_change</th>\n",
       "      <th>sloop_entropy_wt</th>\n",
       "      <th>solvation_hydrophobic_change</th>\n",
       "      <th>solvation_hydrophobic_wt</th>\n",
       "      <th>solvation_polar_change</th>\n",
       "      <th>solvation_polar_wt</th>\n",
       "      <th>solvent_accessibility_change</th>\n",
       "      <th>solvent_accessibility_wt</th>\n",
       "      <th>t_date_modified</th>\n",
       "      <th>template_errors</th>\n",
       "      <th>torsional_clash_change</th>\n",
       "      <th>torsional_clash_wt</th>\n",
       "      <th>uniprot_domain_id</th>\n",
       "      <th>uniprot_domain_id_old</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>uniprot_mutation</th>\n",
       "      <th>uniprot_name</th>\n",
       "      <th>uniprot_sequence</th>\n",
       "      <th>van_der_waals_change</th>\n",
       "      <th>van_der_waals_clashes_change</th>\n",
       "      <th>van_der_waals_clashes_wt</th>\n",
       "      <th>van_der_waals_wt</th>\n",
       "      <th>water_bridge_change</th>\n",
       "      <th>water_bridge_wt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100.0</td>\n",
       "      <td>67:156</td>\n",
       "      <td>P00648_1b27A.aln</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>67:156</td>\n",
       "      <td>-0.11</td>\n",
       "      <td>63.13</td>\n",
       "      <td>-0.02</td>\n",
       "      <td>-55.36</td>\n",
       "      <td>1b27A00</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>protherm</td>\n",
       "      <td>sp</td>\n",
       "      <td>0.829500</td>\n",
       "      <td>0.09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.58</td>\n",
       "      <td>41.52</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48:157</td>\n",
       "      <td>157</td>\n",
       "      <td>48</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-7.04</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.28</td>\n",
       "      <td>151.88</td>\n",
       "      <td>-0.85</td>\n",
       "      <td>63.15</td>\n",
       "      <td>None</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-1.96</td>\n",
       "      <td>2014-12-22 21:05:37</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48:157</td>\n",
       "      <td>None</td>\n",
       "      <td>P00648.B99990001.pdb</td>\n",
       "      <td>P00648_Q151A/MUT_RepairPDB_P00648.B99990001_1.pdb</td>\n",
       "      <td>P00648_Q151A/WT_RepairPDB_P00648.B99990001_1.pdb</td>\n",
       "      <td>2014-08-07 23:06:32</td>\n",
       "      <td>None</td>\n",
       "      <td>Q104A</td>\n",
       "      <td>-1.33777</td>\n",
       "      <td>110.0</td>\n",
       "      <td>Bacillus amyloliquefaciens</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>bacam/P00/64/P00648/Ribonuclease*67-156/</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>Ribonuclease</td>\n",
       "      <td>Ribonuclease</td>\n",
       "      <td>Ribonuclease</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Ribonuclease</td>\n",
       "      <td>-2.160</td>\n",
       "      <td>120.104122279,87.543977591,49.8328050713,24.78...</td>\n",
       "      <td>0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.78</td>\n",
       "      <td>-26.85</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.47</td>\n",
       "      <td>-158.13</td>\n",
       "      <td>-0.80</td>\n",
       "      <td>159.06</td>\n",
       "      <td>-18.0192</td>\n",
       "      <td>67.1770</td>\n",
       "      <td>2015-09-01 05:12:36</td>\n",
       "      <td>None</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>13.70</td>\n",
       "      <td>24447509</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00648</td>\n",
       "      <td>Q151A</td>\n",
       "      <td>RNBR_BACAM</td>\n",
       "      <td>MMKMEGIALKKRLSWISVCLLVLVSAAGMLFSTAAKTETSSHKAHT...</td>\n",
       "      <td>0.43</td>\n",
       "      <td>-0.28</td>\n",
       "      <td>23.57</td>\n",
       "      <td>-120.88</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100.0</td>\n",
       "      <td>67:156</td>\n",
       "      <td>P00648_1b27A.aln</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>67:156</td>\n",
       "      <td>-0.40</td>\n",
       "      <td>63.28</td>\n",
       "      <td>0.08</td>\n",
       "      <td>-56.00</td>\n",
       "      <td>1b27A00</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>protherm</td>\n",
       "      <td>sp</td>\n",
       "      <td>0.244914</td>\n",
       "      <td>1.63</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-3.04</td>\n",
       "      <td>41.46</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48:157</td>\n",
       "      <td>157</td>\n",
       "      <td>48</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.03</td>\n",
       "      <td>-6.92</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>151.98</td>\n",
       "      <td>-1.21</td>\n",
       "      <td>63.42</td>\n",
       "      <td>None</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-1.96</td>\n",
       "      <td>2014-12-22 21:05:37</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48:157</td>\n",
       "      <td>None</td>\n",
       "      <td>P00648.B99990001.pdb</td>\n",
       "      <td>P00648_I156A/MUT_RepairPDB_P00648.B99990001_1.pdb</td>\n",
       "      <td>P00648_I156A/WT_RepairPDB_P00648.B99990001_1.pdb</td>\n",
       "      <td>2014-08-07 23:03:33</td>\n",
       "      <td>None</td>\n",
       "      <td>I109A</td>\n",
       "      <td>-1.33777</td>\n",
       "      <td>110.0</td>\n",
       "      <td>Bacillus amyloliquefaciens</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>bacam/P00/64/P00648/Ribonuclease*67-156/</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>Ribonuclease</td>\n",
       "      <td>Ribonuclease</td>\n",
       "      <td>Ribonuclease</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Ribonuclease</td>\n",
       "      <td>-2.917</td>\n",
       "      <td>120.104122279,87.543977591,49.8328050713,24.78...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-27.66</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.03</td>\n",
       "      <td>-158.81</td>\n",
       "      <td>-1.25</td>\n",
       "      <td>159.87</td>\n",
       "      <td>5.9086</td>\n",
       "      <td>13.6620</td>\n",
       "      <td>2015-09-01 05:12:36</td>\n",
       "      <td>None</td>\n",
       "      <td>-0.55</td>\n",
       "      <td>13.69</td>\n",
       "      <td>24447509</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00648</td>\n",
       "      <td>I156A</td>\n",
       "      <td>RNBR_BACAM</td>\n",
       "      <td>MMKMEGIALKKRLSWISVCLLVLVSAAGMLFSTAAKTETSSHKAHT...</td>\n",
       "      <td>1.73</td>\n",
       "      <td>-4.68</td>\n",
       "      <td>24.91</td>\n",
       "      <td>-121.42</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100.0</td>\n",
       "      <td>67:156</td>\n",
       "      <td>P00648_1b27A.aln</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>67:156</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>63.31</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-55.95</td>\n",
       "      <td>1b27A00</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>protherm</td>\n",
       "      <td>sp</td>\n",
       "      <td>0.602505</td>\n",
       "      <td>3.47</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.03</td>\n",
       "      <td>42.11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48:157</td>\n",
       "      <td>157</td>\n",
       "      <td>48</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-7.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.22</td>\n",
       "      <td>151.97</td>\n",
       "      <td>-0.47</td>\n",
       "      <td>63.32</td>\n",
       "      <td>None</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-1.96</td>\n",
       "      <td>2014-12-22 21:05:37</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48:157</td>\n",
       "      <td>None</td>\n",
       "      <td>P00648.B99990001.pdb</td>\n",
       "      <td>P00648_V57A/MUT_RepairPDB_P00648.B99990001_1.pdb</td>\n",
       "      <td>P00648_V57A/WT_RepairPDB_P00648.B99990001_1.pdb</td>\n",
       "      <td>2014-08-07 07:15:52</td>\n",
       "      <td>None</td>\n",
       "      <td>V10A</td>\n",
       "      <td>-1.33777</td>\n",
       "      <td>110.0</td>\n",
       "      <td>Bacillus amyloliquefaciens</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>bacam/P00/64/P00648/Ribonuclease*67-156/</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>Ribonuclease</td>\n",
       "      <td>Ribonuclease</td>\n",
       "      <td>Ribonuclease</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Ribonuclease</td>\n",
       "      <td>-3.993</td>\n",
       "      <td>120.104122279,87.543977591,49.8328050713,24.78...</td>\n",
       "      <td>0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.02</td>\n",
       "      <td>-27.66</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.09</td>\n",
       "      <td>-158.30</td>\n",
       "      <td>-0.37</td>\n",
       "      <td>159.86</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2015-09-01 05:12:36</td>\n",
       "      <td>None</td>\n",
       "      <td>-0.81</td>\n",
       "      <td>13.66</td>\n",
       "      <td>24447509</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00648</td>\n",
       "      <td>V57A</td>\n",
       "      <td>RNBR_BACAM</td>\n",
       "      <td>MMKMEGIALKKRLSWISVCLLVLVSAAGMLFSTAAKTETSSHKAHT...</td>\n",
       "      <td>1.04</td>\n",
       "      <td>-0.26</td>\n",
       "      <td>24.97</td>\n",
       "      <td>-121.16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>100.0</td>\n",
       "      <td>67:156</td>\n",
       "      <td>P00648_1b27A.aln</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>67:156</td>\n",
       "      <td>-0.18</td>\n",
       "      <td>63.11</td>\n",
       "      <td>0.07</td>\n",
       "      <td>-56.37</td>\n",
       "      <td>1b27A00</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>protherm</td>\n",
       "      <td>sp</td>\n",
       "      <td>1.360210</td>\n",
       "      <td>0.39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.02</td>\n",
       "      <td>41.92</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48:157</td>\n",
       "      <td>157</td>\n",
       "      <td>48</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.46</td>\n",
       "      <td>-7.31</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.38</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.24</td>\n",
       "      <td>152.13</td>\n",
       "      <td>-0.54</td>\n",
       "      <td>63.01</td>\n",
       "      <td>None</td>\n",
       "      <td>0.19</td>\n",
       "      <td>-2.03</td>\n",
       "      <td>2014-12-22 21:05:37</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48:157</td>\n",
       "      <td>None</td>\n",
       "      <td>P00648.B99990001.pdb</td>\n",
       "      <td>P00648_D59A/MUT_RepairPDB_P00648.B99990001_1.pdb</td>\n",
       "      <td>P00648_D59A/WT_RepairPDB_P00648.B99990001_1.pdb</td>\n",
       "      <td>2014-08-07 06:15:26</td>\n",
       "      <td>None</td>\n",
       "      <td>D12A</td>\n",
       "      <td>-1.33777</td>\n",
       "      <td>110.0</td>\n",
       "      <td>Bacillus amyloliquefaciens</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>bacam/P00/64/P00648/Ribonuclease*67-156/</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>Ribonuclease</td>\n",
       "      <td>Ribonuclease</td>\n",
       "      <td>Ribonuclease</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Ribonuclease</td>\n",
       "      <td>-4.217</td>\n",
       "      <td>120.104122279,87.543977591,49.8328050713,24.78...</td>\n",
       "      <td>0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.48</td>\n",
       "      <td>-27.11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.09</td>\n",
       "      <td>-158.13</td>\n",
       "      <td>-0.73</td>\n",
       "      <td>160.00</td>\n",
       "      <td>-4.6990</td>\n",
       "      <td>52.7104</td>\n",
       "      <td>2015-09-01 05:12:36</td>\n",
       "      <td>None</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>13.58</td>\n",
       "      <td>24447509</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00648</td>\n",
       "      <td>D59A</td>\n",
       "      <td>RNBR_BACAM</td>\n",
       "      <td>MMKMEGIALKKRLSWISVCLLVLVSAAGMLFSTAAKTETSSHKAHT...</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.00</td>\n",
       "      <td>24.97</td>\n",
       "      <td>-121.18</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>100.0</td>\n",
       "      <td>67:156</td>\n",
       "      <td>P00648_1b27A.aln</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>67:156</td>\n",
       "      <td>-0.10</td>\n",
       "      <td>63.13</td>\n",
       "      <td>0.15</td>\n",
       "      <td>-55.89</td>\n",
       "      <td>1b27A00</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>protherm</td>\n",
       "      <td>sp</td>\n",
       "      <td>0.438801</td>\n",
       "      <td>3.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.71</td>\n",
       "      <td>41.49</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48:157</td>\n",
       "      <td>157</td>\n",
       "      <td>48</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.03</td>\n",
       "      <td>-6.98</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.50</td>\n",
       "      <td>151.48</td>\n",
       "      <td>-1.27</td>\n",
       "      <td>62.78</td>\n",
       "      <td>None</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-1.98</td>\n",
       "      <td>2014-12-22 21:05:37</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48:157</td>\n",
       "      <td>None</td>\n",
       "      <td>P00648.B99990001.pdb</td>\n",
       "      <td>P00648_Y60A/MUT_RepairPDB_P00648.B99990001_1.pdb</td>\n",
       "      <td>P00648_Y60A/WT_RepairPDB_P00648.B99990001_1.pdb</td>\n",
       "      <td>2014-08-07 05:31:44</td>\n",
       "      <td>None</td>\n",
       "      <td>Y13A</td>\n",
       "      <td>-1.33777</td>\n",
       "      <td>110.0</td>\n",
       "      <td>Bacillus amyloliquefaciens</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>bacam/P00/64/P00648/Ribonuclease*67-156/</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>Ribonuclease</td>\n",
       "      <td>Ribonuclease</td>\n",
       "      <td>Ribonuclease</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Ribonuclease</td>\n",
       "      <td>-8.450</td>\n",
       "      <td>120.104122279,87.543977591,49.8328050713,24.78...</td>\n",
       "      <td>0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-27.55</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.38</td>\n",
       "      <td>-157.41</td>\n",
       "      <td>-0.35</td>\n",
       "      <td>158.39</td>\n",
       "      <td>-23.5427</td>\n",
       "      <td>35.8754</td>\n",
       "      <td>2015-09-01 05:12:36</td>\n",
       "      <td>None</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>13.64</td>\n",
       "      <td>24447509</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P00648</td>\n",
       "      <td>Y60A</td>\n",
       "      <td>RNBR_BACAM</td>\n",
       "      <td>MMKMEGIALKKRLSWISVCLLVLVSAAGMLFSTAAKTETSSHKAHT...</td>\n",
       "      <td>1.36</td>\n",
       "      <td>-0.02</td>\n",
       "      <td>24.98</td>\n",
       "      <td>-120.34</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   alignment_coverage alignment_def alignment_filename  alignment_identity  alignment_score alignment_subdefs  backbone_clash_change  backbone_clash_wt  backbone_hbond_change  backbone_hbond_wt  cath_id chain chain_modeller  cis_bond_change  cis_bond_wt   dataset  db       ddg  ddg_exp  del_class_exp  del_score_exp  dg_change  dg_wt  disulfide_change  disulfide_wt domain_def  domain_end  domain_start  electrostatic_kon_change  electrostatic_kon_wt  electrostatics_change  electrostatics_wt  energy_ionisation_change  energy_ionisation_wt  entropy_complex_change  entropy_complex_wt  entropy_mainchain_change  entropy_mainchain_wt  entropy_sidechain_change  entropy_sidechain_wt gene_name  helix_dipole_change  helix_dipole_wt     m_date_modified  matrix_score  max_seq_identity  mloop_entropy_change  mloop_entropy_wt model_domain_def model_errors        model_filename                                 model_filename_mut                                 model_filename_wt   mut_date_modified  \\\n",
       "0               100.0        67:156   P00648_1b27A.aln               100.0              1.0            67:156                  -0.11              63.13                  -0.02             -55.36  1b27A00     A              A              0.0          0.0  protherm  sp  0.829500     0.09            NaN            NaN      -0.58  41.52               0.0           0.0     48:157         157            48                       0.0                   0.0                   0.00              -7.04                       0.0                  0.36                     0.0                 0.0                     -0.28                151.88                     -0.85                 63.15      None                 0.00            -1.96 2014-12-22 21:05:37          -1.0               NaN                   0.0               0.0           48:157         None  P00648.B99990001.pdb  P00648_Q151A/MUT_RepairPDB_P00648.B99990001_1.pdb  P00648_Q151A/WT_RepairPDB_P00648.B99990001_1.pdb 2014-08-07 23:06:32   \n",
       "1               100.0        67:156   P00648_1b27A.aln               100.0              1.0            67:156                  -0.40              63.28                   0.08             -56.00  1b27A00     A              A              0.0          0.0  protherm  sp  0.244914     1.63            NaN            NaN      -3.04  41.46               0.0           0.0     48:157         157            48                       0.0                   0.0                   0.03              -6.92                       0.0                  0.36                     0.0                 0.0                     -0.23                151.98                     -1.21                 63.42      None                 0.00            -1.96 2014-12-22 21:05:37          -2.0               NaN                   0.0               0.0           48:157         None  P00648.B99990001.pdb  P00648_I156A/MUT_RepairPDB_P00648.B99990001_1.pdb  P00648_I156A/WT_RepairPDB_P00648.B99990001_1.pdb 2014-08-07 23:03:33   \n",
       "2               100.0        67:156   P00648_1b27A.aln               100.0              1.0            67:156                  -0.06              63.31                   0.01             -55.95  1b27A00     A              A              0.0          0.0  protherm  sp  0.602505     3.47            NaN            NaN       1.03  42.11               0.0           0.0     48:157         157            48                       0.0                   0.0                   0.00              -7.00                       0.0                  0.36                     0.0                 0.0                     -0.22                151.97                     -0.47                 63.32      None                 0.00            -1.96 2014-12-22 21:05:37           0.0               NaN                   0.0               0.0           48:157         None  P00648.B99990001.pdb   P00648_V57A/MUT_RepairPDB_P00648.B99990001_1.pdb   P00648_V57A/WT_RepairPDB_P00648.B99990001_1.pdb 2014-08-07 07:15:52   \n",
       "3               100.0        67:156   P00648_1b27A.aln               100.0              1.0            67:156                  -0.18              63.11                   0.07             -56.37  1b27A00     A              A              0.0          0.0  protherm  sp  1.360210     0.39            NaN            NaN       0.02  41.92               0.0           0.0     48:157         157            48                       0.0                   0.0                   0.46              -7.31                       0.0                  0.38                     0.0                 0.0                     -0.24                152.13                     -0.54                 63.01      None                 0.19            -2.03 2014-12-22 21:05:37          -2.0               NaN                   0.0               0.0           48:157         None  P00648.B99990001.pdb   P00648_D59A/MUT_RepairPDB_P00648.B99990001_1.pdb   P00648_D59A/WT_RepairPDB_P00648.B99990001_1.pdb 2014-08-07 06:15:26   \n",
       "4               100.0        67:156   P00648_1b27A.aln               100.0              1.0            67:156                  -0.10              63.13                   0.15             -55.89  1b27A00     A              A              0.0          0.0  protherm  sp  0.438801     3.00            NaN            NaN       1.71  41.49               0.0           0.0     48:157         157            48                       0.0                   0.0                  -0.03              -6.98                       0.0                  0.36                     0.0                 0.0                     -0.50                151.48                     -1.27                 62.78      None                 0.00            -1.98 2014-12-22 21:05:37          -2.0               NaN                   0.0               0.0           48:157         None  P00648.B99990001.pdb   P00648_Y60A/MUT_RepairPDB_P00648.B99990001_1.pdb   P00648_Y60A/WT_RepairPDB_P00648.B99990001_1.pdb 2014-08-07 05:31:44   \n",
       "\n",
       "  mutation_errors mutation_modeller  norm_dope  number_of_residues               organism_name  partial_covalent_bonds_change  partial_covalent_bonds_wt                              path_to_data  pcv_hbond_change  pcv_hbond_self_change  pcv_hbond_self_wt  pcv_hbond_wt  pcv_salt_equal_change  pcv_salt_equal_self_change  pcv_salt_equal_self_wt  pcv_salt_equal_wt  pcv_salt_opposite_change  pcv_salt_opposite_self_change  pcv_salt_opposite_self_wt  pcv_salt_opposite_wt  pcv_vdw_change  pcv_vdw_self_change  pcv_vdw_self_wt  pcv_vdw_wt  pdbfam_idx   pdbfam_name     pfam_clan    pfam_names  protein_existence  protein_name  provean_score                                         sasa_score  secondary_structure_change  secondary_structure_wt  sequence_version  sidechain_hbond_change  sidechain_hbond_wt  sloop_entropy_change  sloop_entropy_wt  solvation_hydrophobic_change  solvation_hydrophobic_wt  solvation_polar_change  solvation_polar_wt  solvent_accessibility_change  solvent_accessibility_wt  \\\n",
       "0            None             Q104A   -1.33777               110.0  Bacillus amyloliquefaciens                            0.0                        0.0  bacam/P00/64/P00648/Ribonuclease*67-156/               0.0                    0.0                0.0           0.0                    0.0                         0.0                     0.0                0.0                       0.0                            0.0                        0.0                   0.0             0.0                  0.0             14.0         0.0           1  Ribonuclease  Ribonuclease  Ribonuclease                1.0  Ribonuclease         -2.160  120.104122279,87.543977591,49.8328050713,24.78...                           0                     6.0               2.0                    0.78              -26.85                   0.0               0.0                          0.47                   -158.13                   -0.80              159.06                      -18.0192                   67.1770   \n",
       "1            None             I109A   -1.33777               110.0  Bacillus amyloliquefaciens                            0.0                        0.0  bacam/P00/64/P00648/Ribonuclease*67-156/               0.0                    0.0                0.0           0.0                    0.0                         0.0                     0.0                0.0                       0.0                            0.0                        0.0                   0.0             0.0                  0.0             36.0         0.0           1  Ribonuclease  Ribonuclease  Ribonuclease                1.0  Ribonuclease         -2.917  120.104122279,87.543977591,49.8328050713,24.78...                           0                     0.0               2.0                    0.00              -27.66                   0.0               0.0                          3.03                   -158.81                   -1.25              159.87                        5.9086                   13.6620   \n",
       "2            None              V10A   -1.33777               110.0  Bacillus amyloliquefaciens                            0.0                        0.0  bacam/P00/64/P00648/Ribonuclease*67-156/               0.0                    0.0                0.0           0.0                    0.0                         0.0                     0.0                0.0                       0.0                            0.0                        0.0                   0.0             0.0                  0.0             54.0         0.0           1  Ribonuclease  Ribonuclease  Ribonuclease                1.0  Ribonuclease         -3.993  120.104122279,87.543977591,49.8328050713,24.78...                           0                     4.0               2.0                    0.02              -27.66                   0.0               0.0                          2.09                   -158.30                   -0.37              159.86                        0.0000                    0.0000   \n",
       "3            None              D12A   -1.33777               110.0  Bacillus amyloliquefaciens                            0.0                        0.0  bacam/P00/64/P00648/Ribonuclease*67-156/               0.0                    0.0                0.0           0.0                    0.0                         0.0                     0.0                0.0                       0.0                            0.0                        1.0                   0.0             0.0                  0.0             17.0         0.0           1  Ribonuclease  Ribonuclease  Ribonuclease                1.0  Ribonuclease         -4.217  120.104122279,87.543977591,49.8328050713,24.78...                           0                     4.0               2.0                    0.48              -27.11                   0.0               0.0                          0.09                   -158.13                   -0.73              160.00                       -4.6990                   52.7104   \n",
       "4            None              Y13A   -1.33777               110.0  Bacillus amyloliquefaciens                            0.0                        0.0  bacam/P00/64/P00648/Ribonuclease*67-156/               0.0                    0.0                0.0           0.0                    0.0                         0.0                     0.0                0.0                       0.0                            0.0                        0.0                   0.0             0.0                  0.0             37.0         0.0           1  Ribonuclease  Ribonuclease  Ribonuclease                1.0  Ribonuclease         -8.450  120.104122279,87.543977591,49.8328050713,24.78...                           0                     4.0               2.0                    0.00              -27.55                   0.0               0.0                          2.38                   -157.41                   -0.35              158.39                      -23.5427                   35.8754   \n",
       "\n",
       "      t_date_modified template_errors  torsional_clash_change  torsional_clash_wt  uniprot_domain_id  uniprot_domain_id_old uniprot_id uniprot_mutation uniprot_name                                   uniprot_sequence  van_der_waals_change  van_der_waals_clashes_change  van_der_waals_clashes_wt  van_der_waals_wt  water_bridge_change  water_bridge_wt  \n",
       "0 2015-09-01 05:12:36            None                   -0.01               13.70           24447509                    NaN     P00648            Q151A   RNBR_BACAM  MMKMEGIALKKRLSWISVCLLVLVSAAGMLFSTAAKTETSSHKAHT...                  0.43                         -0.28                     23.57           -120.88                  0.0              0.0  \n",
       "1 2015-09-01 05:12:36            None                   -0.55               13.69           24447509                    NaN     P00648            I156A   RNBR_BACAM  MMKMEGIALKKRLSWISVCLLVLVSAAGMLFSTAAKTETSSHKAHT...                  1.73                         -4.68                     24.91           -121.42                  0.0              0.0  \n",
       "2 2015-09-01 05:12:36            None                   -0.81               13.66           24447509                    NaN     P00648             V57A   RNBR_BACAM  MMKMEGIALKKRLSWISVCLLVLVSAAGMLFSTAAKTETSSHKAHT...                  1.04                         -0.26                     24.97           -121.16                  0.0              0.0  \n",
       "3 2015-09-01 05:12:36            None                   -0.01               13.58           24447509                    NaN     P00648             D59A   RNBR_BACAM  MMKMEGIALKKRLSWISVCLLVLVSAAGMLFSTAAKTETSSHKAHT...                  0.23                          0.00                     24.97           -121.18                  0.0              0.0  \n",
       "4 2015-09-01 05:12:36            None                   -0.01               13.64           24447509                    NaN     P00648             Y60A   RNBR_BACAM  MMKMEGIALKKRLSWISVCLLVLVSAAGMLFSTAAKTETSSHKAHT...                  1.36                         -0.02                     24.98           -120.34                  0.0              0.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(947059, 115)\n"
     ]
    }
   ],
   "source": [
    "display(DATA_DF.head())\n",
    "print(DATA_DF.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(498708, 115)\n"
     ]
    }
   ],
   "source": [
    "DATA_DF = DATA_DF.drop_duplicates(\n",
    "    subset=['uniprot_id', 'uniprot_domain_id', 'max_seq_identity', 'uniprot_mutation', 'dataset'], \n",
    "    keep='first')\n",
    "print(DATA_DF.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'clinvar',\n",
      " 'cosmic',\n",
      " 'humsavar',\n",
      " 'protherm',\n",
      " 'protherm_diffseqi',\n",
      " 'taipale',\n",
      " 'taipale_diffseqi'}\n"
     ]
    }
   ],
   "source": [
    "pprint(set(DATA_DF['dataset']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add a train / test split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "VALIDATION_DATASETS = [\n",
    "    'taipale',\n",
    "    'taipale_diffseqi',\n",
    "    'humsavar',\n",
    "    'clinvar',\n",
    "    'cosmic',    \n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "code_folding": [
     0
    ],
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Add a '_train' and '_test' columns\n",
    "ls_all = []\n",
    "\n",
    "for dataset in VALIDATION_DATASETS:\n",
    "    ls_train = []\n",
    "    ls_test = []\n",
    "\n",
    "    df = DATA_DF[(DATA_DF['dataset'] == dataset)]\n",
    "    \n",
    "    if df['ddg_exp'].notnull().all():\n",
    "        df = df.sort_values('ddg_exp')\n",
    "        df1 = df[:df.shape[0] // 2]\n",
    "        df2 = df[df.shape[0] // 2:]\n",
    "        assert df.shape[0] == (df1.shape[0] + df2.shape[0])\n",
    "    elif df['del_class_exp'].notnull().all():\n",
    "        df1 = df[df['del_class_exp'] == 0]\n",
    "        df2 = df[df['del_class_exp'] != 0]\n",
    "        assert df.shape[0] == (df1.shape[0] + df2.shape[0])\n",
    "    else:\n",
    "        raise Exception\n",
    "        \n",
    "    df_train = pd.concat([df1[:-df1.shape[0] // 3], df2[:-df2.shape[0] // 3]], ignore_index=True)\n",
    "    df_test = pd.concat([df1[-df1.shape[0] // 3:], df2[-df2.shape[0] // 3:]], ignore_index=True)\n",
    "    \n",
    "    df_train['dataset'] = dataset + '_train'\n",
    "    df_test['dataset'] = dataset + '_test'\n",
    "\n",
    "    assert DATA_DF[(DATA_DF['dataset'] == dataset)].shape[0] == (df_train.shape[0] + df_test.shape[0])\n",
    "    \n",
    "    ls_all.extend([df_train, df_test])\n",
    "\n",
    "df_all = pd.concat(ls_all, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "code_folding": [
     0
    ],
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Make sure that '_train' in one does not appear in '_test' in other\n",
    "train_mutations = set()\n",
    "\n",
    "for dataset in VALIDATION_DATASETS:\n",
    "    \n",
    "    train_mutations |= set(\n",
    "        df_all[df_all['dataset'] == dataset + '_train']\n",
    "        [['uniprot_id', 'uniprot_mutation']]\n",
    "        .apply('.'.join, axis=1))\n",
    "\n",
    "    df_all.loc[(\n",
    "        (df_all['dataset'] == dataset + '_test') &\n",
    "        (df_all[['uniprot_id', 'uniprot_mutation']].apply('.'.join, axis=1).isin(train_mutations))), \n",
    "        'dataset'] = dataset + '_train'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "code_folding": [
     0
    ],
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                  taipale_train\n",
       "1280                taipale_test\n",
       "1920      taipale_diffseqi_train\n",
       "4980       taipale_diffseqi_test\n",
       "6507              humsavar_train\n",
       "28459              humsavar_test\n",
       "39436              clinvar_train\n",
       "85142               clinvar_test\n",
       "107997              cosmic_train\n",
       "358956               cosmic_test\n",
       "Name: dataset, dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "taipale                    1280        640     0.3333\n",
      "taipale_diffseqi           3122       1465     0.3194\n",
      "humsavar                  22186      10743     0.3262\n",
      "clinvar                   48509      20052     0.2925\n",
      "cosmic                   252758     123682     0.3286\n"
     ]
    }
   ],
   "source": [
    "# Print stats\n",
    "display(df_all['dataset'].drop_duplicates())\n",
    "\n",
    "for dataset in VALIDATION_DATASETS:\n",
    "    n_train = df_all[df_all['dataset'] == dataset + '_train'].shape[0]\n",
    "    n_test = df_all[df_all['dataset'] == dataset + '_test'].shape[0]\n",
    "    print('{:20s} {:10d} {:10d} {:10.4f}'.format(dataset, n_train, n_test, n_test / (n_train + n_test)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DATA_DF_TT = pd.concat([DATA_DF, df_all], ignore_index=True)\n",
    "assert DATA_DF_TT.shape[0] == (DATA_DF.shape[0] + df_all.shape[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "DATA_DF.to_pickle(op.join(NOTEBOOK_NAME, 'DATA_DF.pkl'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DATA_DF_TT.to_pickle(op.join(NOTEBOOK_NAME, 'DATA_DF_TT.pkl'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 5.5G\r\n",
      "drwxrwxr-x 2 strokach kimlab 4.0K Jul 10 21:44 \u001b[0m\u001b[01;34m.\u001b[0m/\r\n",
      "drwxrwxr-x 9 strokach kimlab 4.0K Jul 10 22:01 \u001b[01;34m..\u001b[0m/\r\n",
      "-rw------- 1 strokach kimlab  59M Jun 20 10:25 clinvar.tsv.gz\r\n",
      "-rw------- 1 strokach kimlab 281M Jun 20 10:22 cosmic.tsv.gz\r\n",
      "-rw------- 1 strokach kimlab 1.8G Jul 10 22:02 DATA_DF.pkl\r\n",
      "-rw------- 1 strokach kimlab 2.2G Jul 10 22:02 DATA_DF_TT.pkl\r\n",
      "-rw------- 1 strokach kimlab 3.4G Jun 29 17:53 DATA.pkl\r\n"
     ]
    }
   ],
   "source": [
    "ls $NOTEBOOK_NAME -alh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  },
  "latex_envs": {
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 0
  },
  "nav_menu": {},
  "toc": {
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": false,
   "threshold": 6,
   "toc_cell": false,
   "toc_section_display": "block",
   "toc_window_display": true
  },
  "toc_position": {
   "height": "280px",
   "left": "120.984px",
   "right": "20px",
   "top": "89.9844px",
   "width": "230px"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
